src/rules/if_change_then_change.rs

```diff
@@ -11,54 +12,103 @@ use regex::Regex;
 use crate::diagnostic;
 use crate::git;
 
+#[derive(Debug)]
+pub enum ThenChange {
+    RemoteFile(String),
+    RepoFile(PathBuf),
+    MissingIf,
+    MissingThen,
+    None,
+}
+
 #[derive(Debug)]
 pub struct IctcBlock {
     pub path: PathBuf,
     pub begin: i64,
     pub end: i64,
-    pub thenchange: PathBuf,
+    pub thenchange: ThenChange,
 }
 
-lazy_static::lazy_static! {
-    static ref RE_BEGIN: Regex = Regex::new(r"(?i)^\s*(//|#)\s*ifchange\s*$").unwrap();
-    static ref RE_END: Regex = Regex::new(r"(?i)^\s*(//|#)\s*thenchange\s*(.*)$").unwrap();
+impl IctcBlock {
+    fn get_range(&self) -> diagnostic::Range {
+        diagnostic::Range {
+            path: self.path.to_str().unwrap().to_string(),
+            start: diagnostic::Position {
+                line: self.begin as u64,
+                character: 0,
+            },
+            end: diagnostic::Position {
+                line: self.end as u64,
+                character: 0,
+            },
+        }
+    }
+}
 
+lazy_static::lazy_static! {
+    static ref RE_BEGIN: Regex = Regex::new(r"(?i)^\s*(//|#)\s*ifchange(.*)$").unwrap();
+    static ref RE_END: Regex = Regex::new(r"(?i)^\s*(//|#)\s*thenchange(.*)$").unwrap();
 }
 
 pub fn find_ictc_blocks(path: &PathBuf) -> anyhow::Result<Vec<IctcBlock>> {
     let mut blocks: Vec<IctcBlock> = Vec::new();
 
-    let mut ifttt_begin: i64 = -1;
-
     let in_file = File::open(path).with_context(|| format!("failed to open: {:#?}", path))?;
     let in_buf = BufReader::new(in_file);
 
+    let mut block: Option<IctcBlock> = None;
+
     for (i, line) in lines_view(in_buf)
         .context(format!("failed to read lines of text from: {:#?}", path))?
         .iter()
         .enumerate()
         .map(|(i, line)| (i + 1, line))
     {
         if RE_BEGIN.find(line).is_some() {
-            ifttt_begin = i as i64;
-        } else if let Some(end) = RE_END.captures(line) {
-            if ifttt_begin != -1 {
-                let block = IctcBlock {
+            if let Some(mut block_value) = block {
+                // Two if blocks in a row - report problem
+                block_value.end = block_value.begin;
+                block_value.thenchange = ThenChange::MissingThen;
```

Comment:
nit: this seems like it would be better made as a method on the `IctcBlock` type so it's self-describing. E.g.

```suggestion
                block_value.missing_then();
```

this cuts down on the repeated logic below too

Reply:
Oh, I just mean that this logic should belong to the `IctcBlock` struct so that it is reusable and self-describing

```rust
impl IctcBlock {
    fn get_range(&self) -> diagnostic::Range {
        diagnostic::Range {
            path: self.path.to_str().unwrap().to_string(),
            start: diagnostic::Position {
                line: self.begin.unwrap(),
                character: 0,
            },
            end: diagnostic::Position {
                line: self.end.unwrap(),
                character: 0,
            },
        }
    }
    /// Mark as missing `thenchange` to report as an error later.
    fn mark_as_missing_thenchange&mut self) -> () {
        self.end = self.begin;
        self.thenchange = ThenChange::MissingThen;
    }
}

```